### PR TITLE
[FLINK-20496][state backends] RocksDB partitioned index/filters option.

### PR DESCRIPTION
**Pull request by liuyufei9527**
## What is the purpose of the change

Configure partitioned index and filters options according to 'https://github.com/facebook/rocksdb/wiki/Partitioned-Index-Filters'.

## Brief change log

  - Add `state.backend.rocksdb.memory.partitioned-index-filters` option represent enable this feature
  - Add `state.backend.rocksdb.block.metadata-blocksize` config can set partitioned index block size

## Verifying this change

This change added tests and can be verified as follows:

  - Add test `testGetColumnFamilyOptionsWithPartitionedIndex` in `RocksDBResourceContainerTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes )
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by liuyufei9527** [comment]
@flinkbot run azure

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:147]: From RocksDB's current implementation, partitioned index is only compatible with full bloom filter. From my point of view, we should check whether current filter is full instead of block based and set a default full bloom filter if current filter is not `null`.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for your contribution! I think we could separate this PR into two commits:
1. one commit to introduce partitioned index/filter.
1. one commit to introduce settings of metadata block size.

File: docs/_includes/generated/rocksdb_configurable_configuration.html
Comment (Myasuka) [inline docs/_includes/generated/rocksdb_configurable_configuration.html:27]: Need a space between `block.` and `Currently` and remember to rebuild html document.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (liuyufei9527) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:147]: We couldn't know the filter config(bits, useBlockBasedMode) if it's already created. Should we overwrite with a default full bloom filter? It might have not expected behavior.

**Comment by liuyufei9527** [comment]
> Thanks for your contribution! I think we could separate this PR into two commits:
> 
> 1. one commit to introduce partitioned index/filter.
> 2. one commit to introduce settings of metadata block size.

@Myasuka Should I separate into two commits and do a force push?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (Myasuka) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:147]: I think we could overwrite a default full bloom filter if user configured a filter as this is required for partitioned index feature. We could add warning logs for this behavior change. Remember to add a `TODO` so that we could get the `bits` in configured bloom filter in the future once RocksDB bumped version.

**Comment by Myasuka** [comment]
@liuyufei9527 I suggest to separate this PR into two commits.

**Comment by liuyufei9527** [comment]
@Myasuka Thanks for reviewing, I had update a new version, please take another look.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/PredefinedOptions.java
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/PredefinedOptions.java:165]: I think we could add some basic description of changing to full filter here.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBConfigurableOptions.java
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBConfigurableOptions.java:134]: As we use `MemorySize` to wrap this option, we could no longer need to describe it as `in bytes`.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:245]: This static method could be package level.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBSharedResources.java
Comment (Myasuka) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBSharedResources.java:56]: I think this method could be renamed to `isUsingPartitionedIndexFilters`

**Review by Myasuka — COMMENTED** [review]
Sorry for late review after some internal busy time, please consider to refactor your PR according to the comments, and remember to update the web content.

**Comment by liuyufei9527** [comment]
@Myasuka  Should I fix the conflicts since code format have changed?

**Comment by liuyufei9527** [comment]
@flinkbot run azure

**Comment by Myasuka** [comment]
Your CI was broken due to the bad format, please resubmit your PR with correct format with maven spotless plugin.

Since you actually split two commits in one PR, please use git [fixup](https://thoughtbot.com/blog/autosquashing-git-commits) to point out which newer commit is to fix previous one. You could also remain the PR as only two commits with force push.

**Comment by liuyufei9527** [comment]
@flinkbot run azure

**Comment by liuyufei9527** [comment]
@flinkbot run azure

**Comment by liuyufei9527** [comment]
@Myasuka [CI](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=12150&view=results) have passed.

**Review by Myasuka — APPROVED** [review]
LGTM.
@carp84 do you have more comments?

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:237]:
```suggestion
            LOG.info("Existing filter has been overwritten to full filters since partitioned index filters is enabled.");
```

**Review by carp84 — CHANGES_REQUESTED** [review]
@liuyufei9527 Thanks for the contribution, and sorry for the late response, please see my inline comments.

What's more, somehow I cannot make more than one code suggestions in this review, so there're more coming.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:159]:
```suggestion
            if (rocksResources.isUsingPartitionedIndexFilters() && overwriteFilterIfExist(blockBasedTableConfig)) {
                blockBasedTableConfig.setIndexType(IndexType.kTwoLevelIndexSearch);
                blockBasedTableConfig.setPartitionFilters(true);
                blockBasedTableConfig.setPinTopLevelIndexAndFilter(true);
            }
```
This suggestion indicates that full filters is a necessity to enable partitioned index filters, and if the `overwriteFilterIfExist` is changed to always do the overwriting, then it will always return true and the code could be adjusted accordingly.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:227]:
```suggestion
     * Overwrite configured {@link Filter} when partitioned filter is enabled, since partitioned filter only
     * works with full bloom filter, not blocked based.
```

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:229]: Suggest to directly overwrite the filter no matter what its original type is, since only full filters is a necessity to enable partitioned filter, and the `handlesToClose` collection will make sure the old filter will be closed w/o resource leak (only delayed).

What's more, it's hard to handle the case of failing to get filter from table config through reflection, we cannot let it through silently since the unknown filter type might be block based.

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java:285]:
```suggestion
        BloomFilter blockBasedFilter = new BloomFilter();
```

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java:304]:
```suggestion
                        blockBasedTableConfig.setFilter(blockBasedFilter);
```

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java:286]:
```suggestion
        RocksDBOptionsFactory blockBasedBloomFilterOptionFactory =
```

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java:311]:
```suggestion
                        PredefinedOptions.DEFAULT, blockBasedBloomFilterOptionFactory, opaqueResource)) {
```

File: flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java
Comment (carp84) [inline flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainerTest.java:320]:
```suggestion
            assertThat(getFilterFromBlockBasedTableConfig(actual), not(blockBasedFilter));
        }
        assertFalse("Block based filter is left unclosed.", blockBasedFilter.isOwningHandle());
```

**Review by carp84 — CHANGES_REQUESTED** [review]
@liuyufei9527 This is the 2nd-round review, please check and let me know your thoughts. Thanks.

**Comment by carp84** [comment]
And it seems there are some conflicts to resolve, sorry for the inconvenience due to my late review.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/PredefinedOptions.java
Comment (liuyufei9527) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/PredefinedOptions.java:167]: I'll remove this change, filter will be overwrite after if enable partitioned filter.

File: flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
Comment (liuyufei9527) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:229]: I don't sure unused BloomFilter whether it consumes resources, so I close it manually. I'll revert this if it's ok to keep the useless filter.

Comment (liuyufei9527) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:229]: But still need use reflection verify is using filter, because can't access filter_ property directly.

Comment (carp84) [reply flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java:229]: True, it's better not to set new filter if no filter exists at all. Then the suggested changes are:
1. Make `overwriteFilterIfExist` return a `boolean`
2. Let `getFilterFromBlockBasedTableConfig` throw out the reflection exception to distinguish it from the no filter existing case.
3. Make `overwriteFilterIfExist` return `true` if the filter is successfully overwritten or no filter exists, and return `false` if reflection exception thrown (we don't know whether block based filter is in use and have to disable partitioned index filter).

What do you think? @liuyufei9527 

```suggestion
    private boolean overwriteFilterIfExist(BlockBasedTableConfig blockBasedTableConfig) {
```

**Review by carp84 — CHANGES_REQUESTED** [review]

**Comment by carp84** [comment]
Thanks for the update @liuyufei9527 , please check and fix the check style issue, and make sure the pre-commit build here could pass. Thanks.

> 2021-02-24T12:59:19.9459125Z [ERROR] Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.4.2:check (spotless-check) on project flink-statebackend-rocksdb_2.11: The following files had format violations:
2021-02-24T12:59:19.9459962Z [ERROR] src/main/java/org/apache/flink/contrib/streaming/state/RocksDBResourceContainer.java
2021-02-24T12:59:19.9473145Z [ERROR] Run 'mvn spotless:apply' to fix these violations.

**Comment by liuyufei9527** [comment]
@flinkbot run azure

**Comment by liuyufei9527** [comment]
@flinkbot run azure

**Comment by liuyufei9527** [comment]
@flinkbot run azure

**Review by carp84 — APPROVED** [review]
The latest changes LGTM, will merge it in 24 hours if no more comments/concerns from other committers.

cc @StephanEwen @tillrohrmann

**Comment by StephanEwen** [comment]
Thanks, @liuyufei9527 and @carp84 for taking care of this one.
Looks like a good addition 👍

**Pull request closed without merging**
